File: day4/cpp/main.cpp
#include <algorithm>
#include <cstdio>
#include <ctime>
#include <fstream>
#include <vector>
#include <set>
#include <sstream>
#include <string>
#include <tuple>

#define vec std::vector
#define str std::string
#define set std::set
typedef std::tuple<vec<str>, vec<str>> card;

vec<card> readInput(std::ifstream &file)
{
	vec<card> cards;
	str val;
	file >> val;
	while (file >> val) {
		vec<str> wins;
		vec<str> scratch;
		while (file >> val) {
			if (val != "|") {
				wins.push_back(val);
			} else break;
		}
		while (file >> val) {
			if (val != "Card") {
				scratch.push_back(val);
			} else break;
		}
		cards.push_back({ wins, scratch });
	}
	return cards;
}

int partOne(const vec<card> &input)
{
	int sum = 0;
	for (const card &card: input) {
		vec<str> wins = std::get<0>(card);
		vec<str> scratch = std::get<1>(card);
		std::sort(wins.begin(), wins.end());
		std::sort(scratch.begin(), scratch.end());
		vec<str> intersect;
		std::set_intersection(wins.begin(), wins.end(), scratch.begin(), scratch.end(), std::back_inserter(intersect));
		int res = intersect.size();
		if (res > 0) {
			sum += 1 << (res - 1);
		}
	}
	return sum;
}

int partTwo(const vec<card> &input)
{
	vec<int> lucky;
	for (const card &card: input) {
		vec<str> wins = std::get<0>(card);
		vec<str> scratch = std::get<1>(card);
		std::sort(wins.begin(), wins.end());
		std::sort(scratch.begin(), scratch.end());
		vec<str> intersect;
		std::set_intersection(wins.begin(), wins.end(), scratch.begin(), scratch.end(), std::back_inserter(intersect));
		lucky.push_back(intersect.size());
	}
	
	vec<int> cardnum(lucky.size(), 1);
	for (int i = 0; i < cardnum.size(); i++) {
		for (int j = i + 1; j < i + 1 + lucky[i]; j++) {
			cardnum[j] += cardnum[i];
		}
	}

	int sum = 0;
	for (int val : cardnum) {
		sum += val;
	}
	return sum;
}

int main()
{
	std::ifstream file("..\\input.txt");
    struct timespec start, end;

    timespec_get(&start, TIME_UTC);
	const vec<card> input = readInput(file);
    timespec_get(&end, TIME_UTC);
	printf("input read[us] = %d\n", (end.tv_nsec - start.tv_nsec) / 1000);

    timespec_get(&start, TIME_UTC);
	int res1 = partOne(input);
    timespec_get(&end, TIME_UTC);
	printf("part one = %d\n", res1);
	printf("time[us] = %d\n", (end.tv_nsec - start.tv_nsec) / 1000);

    timespec_get(&start, TIME_UTC);
	int res2 = partTwo(input);
    timespec_get(&end, TIME_UTC);
	printf("part one = %d\n", res2);
	printf("time[us] = %d\n", (end.tv_nsec - start.tv_nsec) / 1000);
	return 0;
}

File: day4/cs/main.cs

using System.Diagnostics;

List<string> readNumStr(string line, int start, int end)
{
    List<string> list = new List<string>();
    char[] buff = new char[10];
    int count = 0;

    for (int i = start; i < end; i++)
    {
        if (Char.IsDigit(line[i]))
        {
            buff[count++] = line[i];
        } else if (count > 0)
        {
            list.Add(new string(buff, 0, count));
            count = 0;
        }
    }
    if (count > 0)
    {
        list.Add(new string(buff, 0, count));
    }
    return list;
}

List<string> readNumStr2(string line, int start, int end)
{
    List<string> list = new List<string>();

    string[] nums = line.Substring(start, end - start).Split(' ');
    foreach (string str in nums)
    {
        if (str != "") { list.Add(str); }
    }
    return list;
}
List<Card> readInput(string file)
{
    List<Card> cards = new List<Card>();
    List<string> wins = new List<string>();
    List<string> scratch = new List<string>();

    string line;
    StreamReader sr = new StreamReader(file);
    while ((line = sr.ReadLine()) != null)
    {
        int pos1 = line.IndexOf(':');
        int pos2 = line.IndexOf('|');
        wins = readNumStr2(line, pos1 + 1, pos2);
        scratch = readNumStr2(line, pos2 + 1, line.Length);
        cards.Add(new Card { wins = wins,  scratch = scratch });
    }
    return cards;
}

int partOne(List<Card> cards)
{
    int sum = 0;
    foreach (Card card in cards)
    {
        List<string> intersect = new List<string>(card.scratch.Intersect(card.wins));
        int res = intersect.Count;
        if (res > 0) { sum += 1 << (res - 1); }
    }
    return sum;
}

int partTwo(List<Card> cards)
{    
    List<int> lucky = new List<int>();

    foreach (Card card in cards)
    {
        List<string> intersect = new List<string>(card.scratch.Intersect(card.wins));
        lucky.Add(intersect.Count);
    }

    int[] cardnum = new int[lucky.Count];
    Array.Fill(cardnum, 1);
    for (int i = 0; i < cardnum.Length; i++)
    {
        for (int j = i + 1; j < i + 1 + lucky[i]; j++)
        {
            cardnum[j] += cardnum[i];
        }
    }

    int sum = 0;
    foreach (int val in cardnum)
    {
        sum += val;
    }
    return sum;
}

Stopwatch timer = new Stopwatch();

timer.Start();
List<Card> input = readInput(@"..\..\..\..\input.txt");
timer.Stop();
Console.WriteLine(String.Format("loading time[us] = {0}", timer.Elapsed.TotalMicroseconds));

timer.Restart();
int res1 = partOne(input);
timer.Stop();
Console.WriteLine(String.Format("part one = {0}", res1));
Console.WriteLine(String.Format("time[us] = {0}", timer.Elapsed.TotalMicroseconds));

timer.Restart();
int res2 = partTwo(input);
timer.Stop();
Console.WriteLine(String.Format("part two = {0}", res2));
Console.WriteLine(String.Format("time[us] = {0}", timer.Elapsed.TotalMicroseconds));
struct Card
{
    public List<string> wins;
    public List<string> scratch;
};

